frontend/src/App.tsx

```diff
@@ -1,46 +1,112 @@
-import { useState } from "react";
+import { useCallback, useEffect, useState } from "react";
 import "./App.css";
-import { GetGPTResponse } from "../wailsjs/go/main/App";
+import { GetGPTResponse, GetUserName } from "../wailsjs/go/main/App";
 function App() {
   const [resultText, setResultText] = useState("");
-  const [name, setName] = useState("Brian");
-  const updateName = (e: any) => setName(e.target.value);
+  const [query, setQuery] = useState("");
+  const [userName, setUserName] = useState("dear");
+  const [isGptResponseLoading, setIsGptResponseLoading] = useState(false);
+  const [isTyping, setIsTyping] = useState(false);
 
-  function getGPTResponse() {
-    GetGPTResponse(name).then(typeResponseGradually);
-  }
+  const updateQuery = (e: any) => setQuery(e.target.value);
 
-  function typeResponseGradually(message: string) {
+  const getGPTResponse = useCallback(async () => {
+    if (!query) {
+      setResultText(
+        "Query cannot be empty, please enter something in the search bar"
+      );
+      return;
+    }
+
+    if (isGptResponseLoading || isTyping) {
+      return;
+    }
+
+    setIsGptResponseLoading(true);
+    const gptResponse = await GetGPTResponse(query);
+    setIsGptResponseLoading(false);
+
+    typeResponseGradually(gptResponse);
+  }, [query, isTyping, isGptResponseLoading]);
+
+  const typeResponseGradually = (message: string) => {
     setResultText(""); // Clear any previous text
-    let index = -1;
-    const speed = 50; // Speed of typing effect in milliseconds
+    setIsTyping(true);
+    let index = 0;
+    const speed = 50;
 
-    function typeChar() {
+    const typeChar = () => {
       if (index < message.length) {
-        setResultText((prev) => {
-          return prev + message.charAt(index);
-        });
+        setResultText((prev) => prev + message.charAt(index));
         index++;
-        setTimeout(typeChar, speed); // Recursively call to add the next character
+        setTimeout(typeChar, speed);
+      } else {
+        setIsTyping(false);
+      }
+    };
+
+    typeChar();
+  };
+
+  // Handle keyboard events
+  const handleInputEvent = useCallback(
+    async (event: KeyboardEvent) => {
+      const key = event.key;
+      console.log(event);
+
+      switch (key) {
+        case "Enter":
+          if (isGptResponseLoading || isTyping) {
+            return;
+          }
+          await getGPTResponse();
+          return;
+        case "Escape":
+          console.log("closing");
+          return;
+      }
+    },
+    [query, isGptResponseLoading, isTyping, getGPTResponse]
+  );
+
+  // Fetch username on component mount
+  useEffect(() => {
+    async function fetchUserName() {
```

Comment:
Can we be consistent on usage of `function` vs `const` syntax? Or maybe we can determine which is better to use in what case.

Reply:
True, `const` supremacy